### PR TITLE
[SPARK-31233][Core] Enhance RpcTimeoutException Log Message

### PR DESCRIPTION
**Pull request by Huang-yi-3456**
### What changes were proposed in this pull request?

askAbortable method throw TimeoutException while it does no complete in time. Currently, the error message contains null as remoteAddr when receiver is in client mode. 
This change is to print out correct rpcAddress instead of null in the error message.

### Why are the changes needed?

It provides the address of an endpoint which does not reply in time. It helps users to find slow executors when timeout happens. 

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Add a unit test.

### CODE REVIEW

**Comment by jiangxb1987** [comment]
Please revert the unnecessary code reformat changes.

**Comment by jiangxb1987** [comment]
OK to test

**Comment by Huang-yi-3456** [comment]
Hi @jiangxb1987 ,
it's reverted.

**Comment by jiangxb1987** [comment]
QQ: Under which case would remoteAddr be assigned a null value?

**Comment by Huang-yi-3456** [comment]
@jiangxb1987 
Here is the Comment for class RpcEndpointAddress.
>  * The `rpcAddress` may be null, in which case the endpoint is registered via a client-only
 * connection and can only be reached via the client that sent the endpoint reference.

rpcAddress is init by the following method
`  override lazy val address: RpcAddress = {
    if (server != null) RpcAddress(host, server.getPort()) else null
  }
` and server is only init once it is not client mode
`(if !config.clientMode)`

File: core/src/test/scala/org/apache/spark/rpc/netty/NettyRpcEnvSuite.scala [hunk 1]
```diff
@@ -42,6 +50,34 @@ class NettyRpcEnvSuite extends RpcEnvSuite with MockitoSugar with TimeLimits {
     new NettyRpcEnvFactory().create(config)
   }
 
+  test("SPARK-31233: Send message to clientMode RpcEnv with timeout") {
```
Comment (jiangxb1987) [inline core/src/test/scala/org/apache/spark/rpc/netty/NettyRpcEnvSuite.scala]: maybe i'm missing something, but why not create a RpcEnv with clientMode = true?

Comment (Huang-yi-3456) [reply core/src/test/scala/org/apache/spark/rpc/netty/NettyRpcEnvSuite.scala]: Hi @jiangxb1987 got it. The unit test is updated.

**Comment by dongjoon-hyun** [comment]
ok to test

File: core/src/test/scala/org/apache/spark/rpc/RpcEnvSuite.scala [hunk 2]
```diff
@@ -190,6 +190,47 @@ abstract class RpcEnvSuite extends SparkFunSuite with BeforeAndAfterAll {
...
+          Thread.sleep(1200)
+          context.reply("okay")
+      }
+    })
+    try {
+      val reply = rpcEndpointRef.askSync[String](Register(anotherRef))
+      assert("okay" === reply)
+      val answer = remoteRef.ask[String]("msg",
+        RpcTimeout(conf, Seq("spark.rpc.askTimeout", "spark.network.timeout"), "1s"))
+      val thrown = intercept[Exception] {
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/rpc/RpcEnvSuite.scala:223]: `val thrown = intercept[RpcTimeoutException] {`?

File: core/src/test/scala/org/apache/spark/rpc/RpcEnvSuite.scala [hunk 3]
```diff
@@ -190,6 +190,47 @@ abstract class RpcEnvSuite extends SparkFunSuite with BeforeAndAfterAll {
...
+        ThreadUtils.awaitResult(answer, Duration(1300, MILLISECONDS))
+      }
+      assert(thrown.getClass.equals(classOf[RpcTimeoutException]))
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/rpc/RpcEnvSuite.scala:226]: It seems that you can remove this if you do https://github.com/apache/spark/pull/28002/files#r401946844 .

File: core/src/test/scala/org/apache/spark/rpc/RpcEnvSuite.scala [hunk 4]
```diff
@@ -190,6 +190,47 @@ abstract class RpcEnvSuite extends SparkFunSuite with BeforeAndAfterAll {
...
+      assert(!thrown.getMessage.contains("null"))
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/rpc/RpcEnvSuite.scala:227]: Can we check the exact message?

**Comment by dongjoon-hyun** [comment]
Hi, @Huang-yi-3456 . Did you push your last commit?

**Comment by Huang-yi-3456** [comment]
Hi @dongjoon-hyun ,
Yes I did. It is 9c50b8cdc26cc8d3172a8218d8debd09c3288556 on my branch. But Github for some reason has not synced it here.

File: core/src/test/scala/org/apache/spark/rpc/RpcEnvSuite.scala (hunk 4 above)
Comment (Huang-yi-3456) [reply core/src/test/scala/org/apache/spark/rpc/RpcEnvSuite.scala]: Fixed.

**Comment by Huang-yi-3456** [comment]
@dongjoon-hyun Please kindly review the changes, thanks.

**Comment by Huang-yi-3456** [comment]
@jiangxb1987  @dongjoon-hyun 
Could you please let me know if there is anything else required for this PR? 
Thanks.

File: core/src/test/scala/org/apache/spark/rpc/netty/NettyRpcEnvSuite.scala [hunk 5]
```diff
@@ -135,4 +136,49 @@ class NettyRpcEnvSuite extends RpcEnvSuite with MockitoSugar with TimeLimits {
...
+      val reply = rpcEndpointRef.askSync[String](Register(anotherRef))
+      assert("okay" === reply)
+      val timeout = "1s"
+      val answer = remoteRef.ask[String]("msg",
+        RpcTimeout(conf, Seq("spark.rpc.askTimeout", "spark.network.timeout"), timeout))
+      val thrown = intercept[RpcTimeoutException] {
+        ThreadUtils.awaitResult(answer, Duration(1300, MILLISECONDS))
+      }
+      val remoteAddr = remoteRef.asInstanceOf[NettyRpcEndpointRef].client.getChannel.remoteAddress
+      val exptMsg = s"Cannot receive any reply from $remoteAddr in 1 second. " +
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/rpc/netty/NettyRpcEnvSuite.scala:176]: I wouldn't be this specific in testing error messages. Just assert it contains the address you are expecting and/or maybe another key phrase

Comment (Huang-yi-3456) [reply core/src/test/scala/org/apache/spark/rpc/netty/NettyRpcEnvSuite.scala]: thanks @srowen It is updated.

**Comment by dongjoon-hyun** [comment]
Thank you for updating, @Huang-yi-3456 . The last commit is a string change in test suite and I verified that.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @Huang-yi-3456 , @jiangxb1987 , @srowen .
Merged to master for Apache Spark 3.1.0 as an enhancement.

**Comment by dongjoon-hyun** [comment]
@Huang-yi-3456 . You are added to the Apache Spark contributor group and SPARK-31233 is assigned to you. Thank you again.

**Pull request closed without merging**